File: alljoyn_java/src/org/alljoyn/bus/Observer.java
/*
 * Copyright (c) 2015, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.alljoyn.bus;

import java.util.NavigableMap;
import java.util.TreeMap;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

import org.alljoyn.bus.annotation.BusInterface;

/**
 * A convenient object discovery and access mechanism.
 * <p>
 * An Observer takes care of discovery, session management and ProxyBusObject
 * creation for bus objects that implement a specific set of interfaces.
 * <p>
 * The Observer monitors About announcements, and automatically sets up sessions
 * with all peers that offer objects of interest (i.e. objects that implement at
 * least the set of mandatory interfaces for this Observer). The Observer creates
 * a proxy bus object for each discovered object. The Observer::Listener class
 * is used to inform the application about the discovery of new objects, and about
 * the disappearance of objects from the bus.
 * <p>
 * Objects are considered lost in the following cases:
 * <ul>
 * <li> they are un-announced via About
 * <li> the hosting peer has closed the session
 * <li> the hosting peer stopped responding to Ping requests
 * </ul>
 */
public class Observer {
    /**
     * Listener for object lifecycle events
     */
    public interface Listener {
        /**
         * A new object has been discovered.
         *
         * @param object a proxy for the discovered object, supporting all
         *               mandatory interfaces, and those optional interfaces
         *               that the object implements.
         */
        void ObjectDiscovered(ProxyBusObject object);

        /**
         * A previously discovered object has been lost.
         *
         * The proxy bus object is from this point on invalid. Even if the
         * object reappears on the bus, you'll need a new proxy object (and
         * objectDiscovered will provide you with that new proxy) to interact
         * with the bus object.
         *
         * @param object the proxy for the object that has been lost.
         */
        void ObjectLost(ProxyBusObject object);
    };

    /**
     * Create an Observer that discovers all objects implementing a minimum set
     * of Interfaces.
     *
     * @param  bus  the bus attachment to be used by this Observer
     * @param  mandatoryInterfaces the minimal set of Interfaces that have to
     *                             be implemented by a bus object for the
     *                             Observer to discover that object.
     * @param optionalInterfaces additional Interfaces that will be added to the
     *                           {@link ProxyBusObject}s managed created by this
     *                           Observer, if they are supported by the
     *                           discovered bus objects.
     */
    public Observer(BusAttachment bus, Class<?>[] mandatoryInterfaces, Class<?>[] optionalInterfaces) {
        this.bus = bus;
        proxies = new TreeMap<ObjectId, ProxyBusObject>();
        listeners = new ArrayList<WrappedListener>();
        interfaceMap = new HashMap<String, Class<?>>();

        /* build the list of mandatory AllJoyn interface names */
        String[] mandatoryNames = new String[mandatoryInterfaces.length];
        for (int i = 0; i < mandatoryInterfaces.length; ++i) {
            String name = getBusInterfaceName(mandatoryInterfaces[i]);
            mandatoryNames[i] = name;
            interfaceMap.put(name, mandatoryInterfaces[i]);
        }
        if (optionalInterfaces != null) {
            for (Class<?> intf : optionalInterfaces) {
                interfaceMap.put(getBusInterfaceName(intf), intf);
            }
        }

        create(bus, mandatoryNames);
    }

    /**
     * Create an Observer that discovers all objects implementing a minimum set
     * of Interfaces.
     *
     * @param  bus  the bus attachment to be used by this Observer
     * @param  mandatoryInterfaces the minimal set of Interfaces that have to
     *                             be implemented by a bus object for the
     *                             Observer to discover that object.
     */
    public Observer(BusAttachment bus, Class<?>[] mandatoryInterfaces) {
        this(bus, mandatoryInterfaces, new Class<?>[] {});
    }

    /**
     * Register an {@link Observer.Listener}.
     *
     * @param listener the listener to register
     * @param triggerOnExisting indicates whether the listener's
     *                          objectDiscovered callback should be invoked for
     *                          all objects that have been discovered prior to
     *                          the registration of this listener.
     */
    public synchronized void registerListener(Listener listener, boolean triggerOnExisting) {
        listeners.add(new WrappedListener(listener, !triggerOnExisting));
        if (triggerOnExisting) {
            triggerEnablePendingListeners();
        }
    }

    /**
     * Register an {@link Observer.Listener}.
     *
     * The listener's objectDiscovered callback will be invoked for all objects
     * that have been discovered prior to the registration of the listener.
     *
     * @param listener the listener to register
     */
    public synchronized void registerListener(Listener listener) {
        registerListener(listener, true);
    }

    /**
     * Unregister a previously registered {@link Observer.Listener}.
     *
     * @param listener the listener to unregister
     */
    public synchronized void unregisterListener(Listener listener) {
        for (int i = 0; i < listeners.size(); ++i) {
            if (listeners.get(i).listener == listener) {
                listeners.remove(i);
                break;
            }
        }
    }

    /**
     * Unregister all listeners.
     */
    public synchronized void unregisterAllListeners() {
        listeners.clear();
    }

    /**
     * Retrieve a {@link ProxyBusObject}.
     *
     * If the supplied (busname, path) pair does not identify an object that
     * has been discovered by this Observer, or identifies an object that has
     * since disappeared from the bus, null will be returned.
     *
     * @param busname  unique bus name of the peer hosting the bus object
     * @param path the object's path
     * @return the ProxyBusObject or null if not found
     */
    public synchronized ProxyBusObject get(String busname, String path) {
        ObjectId oid = new ObjectId(busname, path);
        return proxies.get(oid);
    }

    /**
     * Retrieve the first {@link ProxyBusObject}.
     *
     * The getFirst/getNext pair of methods is used for iteration over all
     * discovered objects. null is returned to signal end of iteration.
     *
     * @return the first ProxyBusObject or null if no objects discovered
     */
    public synchronized ProxyBusObject getFirst() {
        if (proxies.size() == 0) {
            return null;
        }
        return proxies.firstEntry().getValue();
    }

    /**
     * Retrieve the next {@link ProxyBusObject}.
     *
     * The getFirst/getNext pair of methods is used for iteration over all
     * discovered objects. null is returned to signal end of iteration.
     *
     * @param object the previous ProxyBusObject
     * @return the next ProxyBusObject or null if iteration is finished
     */
    public synchronized ProxyBusObject getNext(ProxyBusObject object) {
        if (object == null) {
            return null;
        }

        Map.Entry<ObjectId, ProxyBusObject> entry = proxies.ceilingEntry(new ObjectId(object));
        if (entry == null) {
            return null;
        }

        return entry.getValue();
    }

    /**
     * Let the Java garbage collector release resources.
     */
    @Override
    protected void finalize() throws Throwable {
        try {
            destroy();
        } finally {
            super.finalize();
        }
    }

    /**
     * Points to the underlying native JObserver
     */
    private long handle;

    /**
     * Create native counterpart.
     */
    private native void create(BusAttachment bus, String[] interfaceNames);

    /**
     * Destroy native counterpart.
     */
    private synchronized native void destroy();

    /**
     * Make native call us back on the dispatcher thread.
     */
    private native void triggerEnablePendingListeners();

    /**
     * Encapsulates the unique id of a bus object (the busname/path pair).
     */
    private static class ObjectId implements Comparable<ObjectId> {
        /**
         * unique bus name
         */
        public String busname;

        /**
         * object path
         */
        public String path;

        ObjectId(String _busname, String _path) {
            busname = _busname;
            path = _path;
        }

        ObjectId(ProxyBusObject proxy) {
            this(proxy.getBusName(), proxy.getObjPath());
        }

        @Override
        public int hashCode() {
            return busname.hashCode() ^ path.hashCode();
        }

        @Override
        public boolean equals(Object other) {
            if (other == this) {
                return true;
            }
            if (other == null) {
                return false;
            }
            if (other.getClass() != getClass()) {
                return false;
            }

            ObjectId oid = (ObjectId) other;
            return busname.equals(oid.busname) && path.equals(oid.path);
        }

        @Override
        public int compareTo(ObjectId oid) {
            if (oid == null) {
                return 1;
            }

            int nameCmp = busname.compareTo(oid.busname);
            if (nameCmp == 0) {
                return path.compareTo(oid.path);
            }
            return nameCmp;
        }
    };

    /**
     * Wraps a listener to add some bookkeeping information.
     */
    private static class WrappedListener {
        /**
         * Is the listener enabled?
         *
         * triggerOnExisting listeners start out disabled, they get enabled
         * when native calls us back on the dispatcher thread via
         * enablePendingListeners.
         */
        public boolean enabled;

        /**
         * the listener
         */
        public Listener listener;

        WrappedListener(Listener l, boolean enable) {
            enabled = enable;
            listener = l;
        }
    };

    /**
     * Enable pending listeners.
     *
     * Listeners that are registered with triggerOnExisting == true, start off
     * in pending state. That means we don't dispatch callbacks to them until
     * we've been able to invoke objectDiscovered for all objects that are
     * already in our proxy list. Native calls us via this method to do those
     * initial callbacks.
     */
    private void enablePendingListeners() {
        ArrayList<Listener> pendingListeners = new ArrayList<Listener>();
        synchronized(this) {
            for (WrappedListener wl : listeners) {
                if (!wl.enabled) {
                    pendingListeners.add(wl.listener);
                    wl.enabled = true;
                }
            }
        }

        /* we have threading guarantees from the Core that the proxies set will not change
         * during this call. Therefore, we don't need to do anything special to safeguard
         * proxies during iteration. */
        for (Listener l : pendingListeners) {
            for (ProxyBusObject proxy : proxies.values()) {
                l.ObjectDiscovered(proxy);
            }
        }
    }

    /**
     * Native tells us a new object was discovered.
     *
     * @param busname unique bus name
     * @param path the object path
     * @param interfaces list of announced AllJoyn Interfaces for the discovered object
     * @param sessionId id of the session we can use to construct a ProxyBusObject
     */
    private void objectDiscovered(String busname, String path, String[] interfaces, int sessionId) {
        ArrayList<Class<?>> intfList = new ArrayList<Class<?>>();
        for (String intfName : interfaces) {
            Class<?> intf = interfaceMap.get(intfName);
            if (intf != null) {
                intfList.add(intf);
            }
        }
        //TODO figure out what to do with secure bus objects
        ProxyBusObject proxy = bus.getProxyBusObject(busname, path, sessionId, intfList.toArray(new Class<?>[]{}));

        WrappedListener[] copiedListeners;
        synchronized(this) {
            proxies.put(new ObjectId(busname, path), proxy);
            copiedListeners = listeners.toArray(new WrappedListener[]{});
        }

        /* we do the listener invocation outside of the critical region to avoid
         * lock ordering issues */
        for (WrappedListener wl : copiedListeners) {
            if (wl.enabled) {
                wl.listener.ObjectDiscovered(proxy);
            }
        }
    }

    /**
     * Native tells us an object has disappeared.
     *
     * @param busname unique bus name
     * @param path object path
     */
    private void objectLost(String busname, String path) {
        WrappedListener[] copiedListeners = null;
        ProxyBusObject obj = null;
        synchronized(this) {
            ObjectId oid = new ObjectId(busname, path);
            obj = proxies.remove(oid);
            if (obj != null) {
                copiedListeners = listeners.toArray(new WrappedListener[]{});
            }
        }

        if (obj == null) {
            return;
        }

        /* we do the listener invocation outside of the critical region to avoid
         * lock ordering issues */
        for (WrappedListener wl : copiedListeners) {
            if (wl.enabled) {
                wl.listener.ObjectLost(obj);
            }
        }
    }

    /**
     * Extract AllJoyn interface name from a Java interface.
     */
    private static String getBusInterfaceName(Class<?> intf) {
         BusInterface annotation = intf.getAnnotation(BusInterface.class);
         if (annotation != null && !annotation.name().equals("")) {
             return annotation.name();
         }
         return intf.getCanonicalName();
    }

    /** Bus attachment used by this Observer */
    private BusAttachment bus;
    /** Proxies for the discovered objects */
    private NavigableMap<ObjectId, ProxyBusObject> proxies;
    /** Registered listeners */
    private List<WrappedListener> listeners;
    /** Maps AllJoyn interface names to Java interfaces (needed for ProxyBusObject creation) */
    private HashMap<String, Class<?>> interfaceMap;
}

File: alljoyn_core/src/Observer.cc
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <set>

#include <qcc/String.h>
#include <qcc/Mutex.h>

#include <alljoyn/Observer.h>
#include <alljoyn/InterfaceDescription.h>

#include "BusInternal.h"
#include "CoreObserver.h"
#include "ObserverManager.h"
#include "BusUtil.h"

#include <qcc/Debug.h>
#define QCC_MODULE "OBSERVER"

using namespace std;

namespace ajn {

ObjectId::ObjectId() {
}

ObjectId::ObjectId(const qcc::String& busname, const qcc::String& path) :
    uniqueBusName(IsLegalUniqueName(busname.c_str()) ? busname : ""),
    objectPath(IsLegalObjectPath(path.c_str()) ? path : "") {
}

ObjectId::ObjectId(const ManagedProxyBusObject& mpbo) :
    uniqueBusName(mpbo->GetUniqueName()), objectPath(mpbo->GetPath()) {
}

ObjectId::ObjectId(const ProxyBusObject* ppbo) :
    uniqueBusName((ppbo != NULL ? ppbo->GetUniqueName() : "")),
    objectPath((ppbo != NULL ? ppbo->GetPath() : "")) {
}

ObjectId::ObjectId(const ProxyBusObject& pbo) :
    uniqueBusName(pbo.GetUniqueName()), objectPath(pbo.GetPath()) {
}

ObjectId::ObjectId(const ObjectId& other) :
    uniqueBusName(other.uniqueBusName), objectPath(other.objectPath) {
}

ObjectId& ObjectId::operator=(const ObjectId& other) {
    uniqueBusName = other.uniqueBusName;
    objectPath = other.objectPath;
    return *this;
}

bool ObjectId::operator==(const ObjectId& other) const {
    return uniqueBusName == other.uniqueBusName && objectPath == other.objectPath;
}

bool ObjectId::operator<(const ObjectId& other) const {
    return (uniqueBusName == other.uniqueBusName) ? (objectPath < other.objectPath) :
           (uniqueBusName < other.uniqueBusName);
}

bool ObjectId::IsValid() const {
    return (uniqueBusName != "") && (objectPath != "");
}

class Observer::Internal : public CoreObserver {
  private:
    BusAttachment& bus;
    Observer* observer;

    /* proxy object bookkeeping */
    typedef std::map<ObjectId, ManagedProxyBusObject> ObjectMap;
    ObjectMap proxies;
    qcc::Mutex proxiesLock;

    /* listener bookkeeping */
    struct WrappedListener {
        /* WrappedListener exists to keep track of whether a given listener
         * is already enabled. triggerOnExisting listeners start off as
         * disabled, until the ObserverManager has the chance to fire the
         * initial callbacks (for "existing" objects) from the work queue. */
        Observer::Listener* listener;
        bool enabled;
        WrappedListener(Observer::Listener* listener, bool enabled)
            : listener(listener), enabled(enabled) { }
    };
    typedef qcc::ManagedObj<WrappedListener*> ProtectedObserverListener;
    typedef std::set<ProtectedObserverListener> ObserverListenerSet;
    ObserverListenerSet listeners;
    qcc::Mutex listenersLock;

  public:
    Internal(BusAttachment& bus,
             Observer* observer,
             InterfaceSet mandatory);

    virtual ~Internal();

    /**
     * Detach from the publicly visible Observer.
     *
     * Because of complicated threading/locking issues, it is not
     * straightforward to just destroy the Internal object from the
     * Observer destructor. Instead, we have a two-phase approach:
     * ~Observer detaches from Internal, and when it is safe to do
     * so, ObserverManager destroys the Internal object.
     */
    void Detach();

    /* implementation of public Observer functionality */
    void RegisterListener(Observer::Listener& listener, bool triggerOnExisting);
    void UnregisterListener(Observer::Listener& listener);
    void UnregisterAllListeners();

    ManagedProxyBusObject Get(const ObjectId& oid);
    ManagedProxyBusObject GetFirst();
    ManagedProxyBusObject GetNext(const ObjectId& oid);

    /* interface towards ObserverManager */
    void ObjectDiscovered(const ObjectId& oid, const std::set<qcc::String>& interfaces, SessionId sessionid);
    void ObjectLost(const ObjectId& oid);
    /**
     * Enable all disabled listeners for this observer.
     *
     * Called from the ObserverManager work queue to make sure the
     * initial callbacks of triggerOnExisting listeners are called
     * from the local endpoint dispatcher threads.
     */
    void EnablePendingListeners();
};

Observer::Internal::Internal(BusAttachment& bus,
                             Observer* observer,
                             InterfaceSet mandatory) :
    CoreObserver(mandatory),
    bus(bus),
    observer(observer)
{
    ObserverManager& obsmgr = bus.GetInternal().GetObserverManager();
    obsmgr.RegisterObserver(this);
}

Observer::Internal::~Internal()
{
}

void Observer::Internal::Detach()
{
    UnregisterAllListeners();
    observer = NULL;

    ObserverManager& obsmgr = bus.GetInternal().GetObserverManager();
    obsmgr.UnregisterObserver(this);
}

void Observer::Internal::RegisterListener(Observer::Listener& listener, bool triggerOnExisting)
{
    Observer::Listener* pListener = &listener;
    WrappedListener* wListener = new WrappedListener(pListener, !triggerOnExisting);
    ProtectedObserverListener protectedListener(wListener);
    listenersLock.Lock(MUTEX_CONTEXT);
    listeners.insert(protectedListener);
    listenersLock.Unlock(MUTEX_CONTEXT);

    if (triggerOnExisting) {
        /* We don't want to do the callbacks from this thread, as it's
         * most probably an application thread. Let the ObserverManager
         * do this for us from the dispatcher thread. To avoid confusion,
         * we leave this listener disabled so that other announcements on
         * the work queue don't get reported out of order.
         */
        ObserverManager& obsmgr = bus.GetInternal().GetObserverManager();
        obsmgr.EnablePendingListeners(this);
    }
}

void Observer::Internal::UnregisterListener(Observer::Listener& listener)
{
    listenersLock.Lock(MUTEX_CONTEXT);

    /* Look for listener on ListenerSet */
    ObserverListenerSet::iterator it;
    for (it = listeners.begin(); it != listeners.end(); ++it) {
        if ((**it)->listener == &listener) {
            break;
        }
    }

    /* Wait for all refs to this ProtectedObserverListener to exit */
    while ((it != listeners.end()) && (it->GetRefCount() > 1)) {
        ProtectedObserverListener l = *it;
        listenersLock.Unlock(MUTEX_CONTEXT);
        qcc::Sleep(5);
        listenersLock.Lock(MUTEX_CONTEXT);
        it = listeners.find(l);
    }

    /* Delete the listeners entry */
    if (it != listeners.end()) {
        ProtectedObserverListener l = *it;
        listeners.erase(it);
        delete (*l);
    }
    listenersLock.Unlock(MUTEX_CONTEXT);
}

void Observer::Internal::EnablePendingListeners()
{
    /* make a list of all pending listeners */
    vector<ProtectedObserverListener> pendingListeners;
    listenersLock.Lock(MUTEX_CONTEXT);
    ObserverListenerSet::iterator lit;
    for (lit = listeners.begin(); lit != listeners.end(); ++lit) {
        ProtectedObserverListener pol = *lit;
        if ((*pol)->enabled == false) {
            pendingListeners.push_back(pol);
        }
    }
    listenersLock.Unlock(MUTEX_CONTEXT);

    /* for each pending listener, invoke ObjectDiscovered for all known proxies */
    proxiesLock.Lock(MUTEX_CONTEXT);
    vector<ProtectedObserverListener>::iterator plit;
    for (plit = pendingListeners.begin(); plit != pendingListeners.end(); ++plit) {
        ProtectedObserverListener pol = *plit;
        (*pol)->enabled = true;

        ObjectMap::iterator it = proxies.begin();
        while (it != proxies.end()) {
            ObjectId id = it->first;
            ManagedProxyBusObject proxy = it->second;
            proxiesLock.Unlock(MUTEX_CONTEXT);
            (*pol)->listener->ObjectDiscovered(proxy);
            proxiesLock.Lock(MUTEX_CONTEXT);
            it = proxies.upper_bound(id);
        }

    }

    proxiesLock.Unlock(MUTEX_CONTEXT);
}

void Observer::Internal::UnregisterAllListeners()
{
    listenersLock.Lock(MUTEX_CONTEXT);

    /* Look for listener on ListenerSet */
    ObserverListenerSet::iterator it = listeners.begin();
    while (it != listeners.end()) {
        /* Wait for all refs to ProtectedBusListener to exit */
        while ((it != listeners.end()) && (it->GetRefCount() > 1)) {
            ProtectedObserverListener l = *it;
            listenersLock.Unlock(MUTEX_CONTEXT);
            qcc::Sleep(5);
            listenersLock.Lock(MUTEX_CONTEXT);
            it = listeners.find(l);
        }

        /* Delete the listeners entry */
        if (it != listeners.end()) {
            ProtectedObserverListener l = *it;
            listeners.erase(it);
            delete (*l);
        }
        it = listeners.begin();
    }
    listenersLock.Unlock(MUTEX_CONTEXT);
}

ManagedProxyBusObject Observer::Internal::Get(const ObjectId& oid)
{
    ManagedProxyBusObject obj;
    if (!oid.IsValid()) {
        return obj;
    }
    proxiesLock.Lock(MUTEX_CONTEXT);
    ObjectMap::iterator it = proxies.find(oid);
    if (it != proxies.end()) {
        obj = it->second;
    }
    proxiesLock.Unlock(MUTEX_CONTEXT);
    return obj;
}

ManagedProxyBusObject Observer::Internal::GetFirst()
{
    ManagedProxyBusObject obj;
    proxiesLock.Lock(MUTEX_CONTEXT);
    ObjectMap::iterator it = proxies.begin();
    if (it != proxies.end()) {
        obj = it->second;
    }
    proxiesLock.Unlock(MUTEX_CONTEXT);
    return obj;
}

ManagedProxyBusObject Observer::Internal::GetNext(const ObjectId& oid)
{
    ManagedProxyBusObject obj;
    if (!oid.IsValid()) {
        return obj;
    }
    proxiesLock.Lock(MUTEX_CONTEXT);
    ObjectMap::iterator it = proxies.upper_bound(oid);
    if (it != proxies.end()) {
        obj = it->second;
    }
    proxiesLock.Unlock(MUTEX_CONTEXT);
    return obj;
}

void Observer::Internal::ObjectDiscovered(const ObjectId& oid,
                                          const std::set<qcc::String>& interfaces,
                                          SessionId sessionid)
{
    /* create proxy object */
    //TODO figure out what to do with secure bus objects
    const char* busname = oid.uniqueBusName.c_str();
    const char* path = oid.objectPath.c_str();
    QCC_DbgTrace(("ObjectDiscovered(%s:%s)", busname, path));

    ManagedProxyBusObject proxy(bus, busname, path, sessionid);
    InterfaceSet::iterator it;
    for (it = interfaces.begin(); it != interfaces.end(); ++it) {
        proxy->AddInterface(it->c_str());
    }

    /* insert in proxy map */
    proxiesLock.Lock(MUTEX_CONTEXT);
    proxies[oid] = proxy;
    proxiesLock.Unlock(MUTEX_CONTEXT);

    /* alert listeners */
    listenersLock.Lock(MUTEX_CONTEXT);
    ObserverListenerSet::iterator lit = listeners.begin();
    while (lit != listeners.end()) {
        ProtectedObserverListener pol = *lit;
        if ((*pol)->enabled == false) {
            ++lit;
            continue;
        }
        listenersLock.Unlock(MUTEX_CONTEXT);
        (*pol)->listener->ObjectDiscovered(proxy);
        listenersLock.Lock(MUTEX_CONTEXT);
        lit = listeners.upper_bound(pol);
    }
    listenersLock.Unlock(MUTEX_CONTEXT);
}

void Observer::Internal::ObjectLost(const ObjectId& oid)
{
    /* remove from proxy map */
    bool found = false;
    ManagedProxyBusObject proxy;

    proxiesLock.Lock(MUTEX_CONTEXT);
    ObjectMap::iterator it = proxies.find(oid);
    if (it != proxies.end()) {
        found = true;
        proxy = it->second;
        proxies.erase(it);
    }
    proxiesLock.Unlock(MUTEX_CONTEXT);

    /* alert listeners */
    if (found) {
        listenersLock.Lock(MUTEX_CONTEXT);
        ObserverListenerSet::iterator lit = listeners.begin();
        while (lit != listeners.end()) {
            ProtectedObserverListener pol = *lit;
            if ((*pol)->enabled == false) {
                ++lit;
                continue;
            }
            listenersLock.Unlock(MUTEX_CONTEXT);
            (*pol)->listener->ObjectLost(proxy);
            listenersLock.Lock(MUTEX_CONTEXT);
            lit = listeners.upper_bound(pol);
        }
        listenersLock.Unlock(MUTEX_CONTEXT);
    }
}

Observer::Observer(BusAttachment& bus,
                   const char* mandatoryInterfaces[],
                   size_t numMandatoryInterfaces)
{
    set<qcc::String> mandatory;
    bool inError = false;

    if (mandatoryInterfaces) {
        for (size_t i = 0; i < numMandatoryInterfaces; ++i) {
            const InterfaceDescription* intf = bus.GetInterface(mandatoryInterfaces[i]);
            if (NULL == intf) {
                QCC_LogError(ER_FAIL, ("Interface %s does not exist", mandatoryInterfaces[i]));
                inError = true;
            } else {
                mandatory.insert(mandatoryInterfaces[i]);
            }
        }
    }

    if (mandatory.empty()) {
        QCC_LogError(ER_FAIL, ("There must be at least one mandatory interface."));
    }

    if (inError) {
        internal = NULL;
    } else {
        internal = new Internal(bus, this, mandatory);
    }
}

Observer::~Observer()
{
    if (internal) {
        internal->Detach();
        internal = NULL;
    }
}

void Observer::RegisterListener(Listener& listener, bool triggerOnExisting)
{
    if (!internal) {
        return;
    }
    internal->RegisterListener(listener, triggerOnExisting);
}

void Observer::UnregisterListener(Listener& listener)
{
    if (!internal) {
        return;
    }
    internal->UnregisterListener(listener);
}

void Observer::UnregisterAllListeners()
{
    if (!internal) {
        return;
    }
    internal->UnregisterAllListeners();
}

ManagedProxyBusObject Observer::Get(const ObjectId& oid)
{
    if (!internal) {
        return ManagedProxyBusObject();
    }
    return internal->Get(oid);
}

ManagedProxyBusObject Observer::GetFirst()
{
    if (!internal) {
        return ManagedProxyBusObject();
    }
    return internal->GetFirst();
}

ManagedProxyBusObject Observer::GetNext(const ObjectId& oid)
{
    if (!internal) {
        return ManagedProxyBusObject();
    }
    return internal->GetNext(oid);
}

}
